### ISSUE TITLE
Fibre Channel volume manager does not find FC disk if multiple targetsWWNs are passed

### ISSUE THREAD

**Issue by zaa** [kubernetes/kubernetes#33602]
**Kubernetes version** (use `kubectl version`): v1.3.5

**Environment**:
- **Cloud provider or hardware configuration**: baremetal
- **OS** (e.g. from /etc/os-release): Fedora 24
- **Kernel** (e.g. `uname -a`): Linux 4.6.6-300.fc24.x86_64 x86_64
- **Install tools**: https://github.com/kubespray/kargo
- **Others**:

**What happened**:

Our kubernetes minions have an FC LUN available via two different WWNs. One set of the servers sees the LUN via target WWN 5001438001340bc4 and the other via 5001438001340bc6. So each server sees the LUN via one WWN (path) only, no DM Multipath is in place.

We would like kubelets to find the LUN via any of the targetWWNs, so we have added the following code into .yaml file:

```
  volumes:
  - name: fc
    fc:
      targetWWNs: ["5001438001340bc4", "5001438001340bc6"]
      lun: 1
```

The node which picked the pod could only connect 5001438001340bc4 WWN.
However we found that kubelet could not find the disk and displayed the following error: "no fc disk found".

**What you expected to happen**:

We expect kubelet to be able to go over all possible targetWWNs and attach the LUN if any of the WWNs can provide the LUN.

**How to reproduce it** (as minimally and precisely as possible):

Place a pod with `fc` volume type with two targetWWNs. Matching WWN should go first. Second WWN can be anything. DM multipath should not be used.

**Anything else do we need to know**:

We have performed an investigation and found out that it appears that the following code https://github.com/kubernetes/kubernetes/blame/master/pkg/volume/fc/fc_util.go#L140 does not handle the case of multiple targetWWNs correctly. For example, if matching WWN comes first and non-matching second, even if the first WWN could be used, the loop will go over to the second WWN and disk var will be empty. Note that we do not use DM multipath, so `dm` device branch does not apply.

We have prepared a patch which should correctly go over all targetWWNs and attach the LUN if any of the WWNs can provide access to the LUN. See the following gist for details:
https://gist.github.com/zaa/b957be41ca37bf1af82b33d47593d9fd

Basically, we changed code to try to find DM multipath-backed disk first. If this fails, we rescan scsi bus and try to use `dm` (if available) **or** plain old `/dev/sd[a-z]` device.

@kubernetes/sig-storage @rootfs

**Comment by zaa** [comment]
Bump

@rootfs @saad-ali

**Comment by rootfs** [comment]
`targetWWNs` expects WWNs that form multpath for the LUN, though I admit this requirement is not explicitly clarified. Non-multipath WWNs are not recommended; it leads to non-deterministic LUN binding to the host.

In your use case, you have to have 2 FC volumes, each volume with its own WWN.

**Comment by zaa** [comment]
@rootfs Thank you for the clarification. Yes, it would be great to have the requirement documented.

**Comment by msau42** [comment]
Looks like this just needs a documentation update?

/sig storage

**Comment by mtanino** [comment]
I made a PR. #3967

https://github.com/kubernetes/kubernetes.github.io/pull/3967

**Comment by mtanino** [comment]
@zaa @rootfs @msau42 
Document was added. I think we can close this issue.
